inject/src/test/groovy/io/micronaut/context/env/PropertySourcePropertyResolverSpec.groovy

```diff
@@ -641,4 +644,79 @@ class PropertySourcePropertyResolverSpec extends Specification {
         then:
         resolver.containsProperty("extra.listval")
     }
+
+    void "test expression resolver"() {
+        given:
+            Map<String, Object> parameters = [foo: "bar"]
+            PropertyResolver mapPropertyResolver = new MapPropertyResolver(parameters)
+            DefaultPropertyPlaceholderResolver propertyPlaceholderResolver = new DefaultPropertyPlaceholderResolver(mapPropertyResolver, ConversionService.SHARED);
+            propertyPlaceholderResolver.@expressionResolvers = [new PropertyExpressionResolver() {
+                @Override
+                @NonNull
+                <T> Optional<T> resolve(@NonNull PropertyResolver propertyResolver,
+                                        @NonNull ConversionService<? extends ConversionService> conversionService,
+                                        @NonNull String expression,
+                                        @NonNull Class<T> requiredType) {
+                    assert requiredType == String.class
+                    assert conversionService
+                    if ("foobar" == expression) {
+                        return Optional.of("ABC")
+                    }
+                    if ("xyz" == expression) {
+                        return Optional.of("123")
+                    }
+                    if ("external" == expression) {
+                        return propertyResolver.get("foo", requiredType)
+                    }
+                    return Optional.empty()
+                }
+            }]
+
+        expect:
+            Optional<String> resolved = propertyPlaceholderResolver.resolvePlaceholders(template)
+            if (result) {
+                assert resolved.isPresent()
+                assert resolved.get() == result
+            } else {
+                assert !resolved.isPresent()
+            }
+        where:
+            template              | result
+            'Hello ${foo}!'      | "Hello bar!"
+            'Hello ${foobar}!'   | "Hello ABC!"
+            'Hello ${xyz}!'      | "Hello 123!"
+            'Hello ${external}!' | "Hello bar!"
+            'Hello ${lol}!'      | null
+    }
+
+    void "test expression resolver is closed"() {
+        given:
+            AtomicBoolean closed = new AtomicBoolean()
+            Map<String, Object> parameters = [foo: "bar"]
+            PropertyResolver mapPropertyResolver = new MapPropertyResolver(parameters)
+            DefaultPropertyPlaceholderResolver propertyPlaceholderResolver = new DefaultPropertyPlaceholderResolver(mapPropertyResolver, ConversionService.SHARED);
+            propertyPlaceholderResolver.@expressionResolvers = [new PropertyExpressionResolverAutoCloseable() {
```

Comment:
This test leverages groovy to bypass  'expressionResolvers'  access rights. This may cause issues in the future.